src/plugins/discover/public/application/main/discover_main_route.tsx

```diff
@@ -109,46 +111,48 @@ export function DiscoverMainRoute({
     page: 'app',
     id: savedSearchId || 'new',
   });
+  /**
+   * Helper function to determine when to skip the no data page
+   */
+  const skipNoDataPage = useCallback(
+    async (nextDataView: DataView | undefined) => {
+      try {
+        const isEsqlQuery = isDataSourceType(
+          stateContainer.appState.getState().dataSource,
+          DataSourceType.Esql
+        );
+        const skipDataTest = savedSearchId || isEsqlQuery;
+
+        if (nextDataView || skipDataTest) {
+          if (!isEsqlQuery) {
+            await stateContainer.actions.loadDataViewList();
+          }
+          return true;
+        }
 
-  const checkData = useCallback(async () => {
-    try {
-      if (savedSearchId) {
-        return true; // bypass NoData screen
-      }
-
-      if (isDataSourceType(stateContainer.appState.getState().dataSource, DataSourceType.Esql)) {
+        const [hasUserDataViewValue, hasESDataValue, defaultDataViewExists] = await Promise.all([
+          data.dataViews.hasData.hasUserDataView().catch(() => false),
+          data.dataViews.hasData.hasESData().catch(() => false),
+          data.dataViews.defaultDataViewExists().catch(() => false),
+          await stateContainer.actions.loadDataViewList(),
+        ]);
```

Comment:
This is the main improvement. Execution in parallel is beneficial for loading performance